File: Assets/Animations/GlowAnimation.ts
import Asset from "Assets/Asset";
import Animation, { AnimationAssets, AnimationOptions } from "./Animation";

interface ConstructionOptions extends AnimationOptions {
  readonly glowStrength?: number;
  readonly glowPeriodSeconds?: number; // pulse period
}

export default class GlowAnimation extends Animation {
  private originalColors: Map<string, [number, number, number, number]> = new Map();
  private glowStrength = 2; // blend amount toward the glow color
  private pulsePeriod = 1.5; // seconds per glow pulse (slower pulse)

  constructor(stopCondition: () => boolean, assets: AnimationAssets[], constructionOptions?: ConstructionOptions) {
    super(assets, constructionOptions);
    this.storeOriginalColors(assets);
    this.glowStrength = constructionOptions?.glowStrength ?? this.glowStrength;
    this.pulsePeriod = constructionOptions?.glowPeriodSeconds ?? this.pulsePeriod;
    this.stopAnimationCondition = stopCondition;
  }

  updateAnimation(deltaTime: number): void {
    super.updateAnimation(deltaTime);
    if (!this.isAnimating) {
      this.restoreOriginalColors();
      return;
    }

    this.animElapsed += deltaTime;

    // Pulsing glow using a sine wave; slower via pulsePeriod
    const intensity = 0.5 + 0.5 * Math.sin((this.animElapsed / this.pulsePeriod) * Math.PI * 2);
    const blend = this.glowStrength * intensity; // how strongly we blend toward glowColor
    const glowColor: [number, number, number] = [1, 0.95, 0.6]; // warm gold tint
    const alphaBoost = 0.2 * intensity;

    for (const asset of this.assets) {
      if (asset instanceof Asset) {
        const original = this.originalColors.get(asset.id);
        if (original) {
          const [r, g, b, a] = original;
          // Lerp original color toward glowColor
          asset.color = [
            r * (1 - blend) + glowColor[0] * blend,
            g * (1 - blend) + glowColor[1] * blend,
            b * (1 - blend) + glowColor[2] * blend,
            Math.min(1, a + alphaBoost),
          ];
        }
      }
    }
  }

  private storeOriginalColors(assets: AnimationAssets[]): void {
    for (const asset of assets) {
      if (asset instanceof Asset) {
        this.originalColors.set(asset.id, [...asset.color]);
      }
    }
  }

  private restoreOriginalColors(): void {
    for (const asset of this.assets) {
      if (asset instanceof Asset) {
        const original = this.originalColors.get(asset.id);
        if (original) {
          asset.color = original;
        }
      }
    }
  }
}

File: Assets/Animations/GlowAnimation.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local Map = ____lualib.Map
local __TS__New = ____lualib.__TS__New
local __TS__InstanceOf = ____lualib.__TS__InstanceOf
local ____exports = {}
local ____Asset = require("Assets.Asset")
local Asset = ____Asset.default
local ____Animation = require("Assets.Animations.Animation")
local Animation = ____Animation.default
____exports.default = __TS__Class()
local GlowAnimation = ____exports.default
GlowAnimation.name = "GlowAnimation"
__TS__ClassExtends(GlowAnimation, Animation)
function GlowAnimation.prototype.____constructor(self, gameManager, id, stopCondition, assets, constructionOptions)
    Animation.prototype.____constructor(
        self,
        gameManager,
        id,
        assets,
        constructionOptions
    )
    self.originalColors = __TS__New(Map)
    self.glowStrength = 2
    self.pulsePeriod = 1.5
    self:storeOriginalColors(assets)
    self.glowStrength = constructionOptions and constructionOptions.glowStrength or self.glowStrength
    self.pulsePeriod = constructionOptions and constructionOptions.glowPeriodSeconds or self.pulsePeriod
    self.stopAnimationCondition = stopCondition
end
function GlowAnimation.prototype.updateAnimation(self, deltaTime)
    Animation.prototype.updateAnimation(self, deltaTime)
    if not self.isAnimating then
        self:restoreOriginalColors()
        return
    end
    self.animElapsed = self.animElapsed + deltaTime
    local intensity = 0.5 + 0.5 * math.sin(self.animElapsed / self.pulsePeriod * math.pi * 2)
    local blend = self.glowStrength * intensity
    local glowColor = {1, 0.95, 0.6}
    local alphaBoost = 0.2 * intensity
    for ____, asset in ipairs(self.assets) do
        if __TS__InstanceOf(asset, Asset) then
            local original = self.originalColors:get(asset.id)
            if original then
                local r, g, b, a = unpack(original, 1, 4)
                asset.color = {
                    r * (1 - blend) + glowColor[1] * blend,
                    g * (1 - blend) + glowColor[2] * blend,
                    b * (1 - blend) + glowColor[3] * blend,
                    math.min(1, a + alphaBoost)
                }
            end
        end
    end
end
function GlowAnimation.prototype.storeOriginalColors(self, assets)
    for ____, asset in ipairs(assets) do
        if __TS__InstanceOf(asset, Asset) then
            self.originalColors:set(
                asset.id,
                {unpack(asset.color)}
            )
        end
    end
end
function GlowAnimation.prototype.restoreOriginalColors(self)
    for ____, asset in ipairs(self.assets) do
        if __TS__InstanceOf(asset, Asset) then
            local original = self.originalColors:get(asset.id)
            if original then
                asset.color = original
            end
        end
    end
end
return ____exports
